### PR TITLE
Mlflow+EarlyStopping

### PR DESCRIPTION
**Pull request by Anya497**
- Create EarlyStopping and add it to the training pipeline.
- Separate two types of learning with study's URI in config file. If you want to train model with concrete hyperparameters you need to specify URI of optuna's study that contains information about the best trial with hyperparameters you interested in. Also you need to add to config model's weights URI which correspond to chosen model's parameters. Otherwise optuna's hyperparameters tuning will be started.
- Add validation type to config. Move number of servers to configuration of validation with SVMs.
- Speed up loss validation with batches.
- Add logging with MLFlow.

### CODE REVIEW

File: AIAgent/ml/training/validation.py [hunk 1]
```diff
@@ -52,21 +55,22 @@ def validate_coverage(
...
+    epoch : int
+        Epoch's number.
     server_count: int
         The number of game servers running in parallel.
     progress_bar_colour : str
         Your favorite colour for progress bar.
     """
     wrapper = TrainingModelWrapper(model)
     tasks = [([game_map], dataset, wrapper) for game_map in dataset.maps]
+    statistics_collector = StatisticsCollector(CURRENT_TABLE_PATH)
```
Comment (Parzival-05) [inline AIAgent/ml/training/validation.py:71]: А нам не нужны красивые таблички с инфой по картам на каждой эпохе или все же это лишнее?

File: AIAgent/ml/training/validation.py [hunk 2]
```diff
@@ -78,38 +82,42 @@ def validate_coverage(
...
-    average_result = np.average(
-        list(
-            map(
-                lambda map_result: map_result.game_result.actual_coverage_percent,
-                all_results,
-            )
+
+    def avg_coverage(results, path_to_coverage: str) -> int:
+        coverage = np.average(
+            list(map(lambda result: getattr(result, path_to_coverage), results))
```
Comment (Parzival-05) [inline AIAgent/ml/training/validation.py:88]: all_results?

File: AIAgent/ml/training/validation.py [hunk 3]
```diff
@@ -78,38 +82,42 @@ def validate_coverage(
...
     )
     statistics_collector.update_results(average_result, all_results)
-    return average_result
+    mlflow.log_metrics(
+        {
+            "average_dataset_state_result": avg_coverage(
+                dataset.maps_results.values(), "coverage_percent"
+            ),
+            "average_result": average_result,
+        },
```
Comment (Parzival-05) [inline AIAgent/ml/training/validation.py:103]: Хорошо бы еще, наверное, прокидывать инфу по каждой SVM? Это, наверное, связано с [этим](https://github.com/gsvgit/PySymGym/pull/75/files#r1688675093)

File: AIAgent/ml/training/validation.py [hunk 4]
```diff
@@ -78,38 +82,42 @@ def validate_coverage(
...
-                all_results,
-            )
+
+    def avg_coverage(results, path_to_coverage: str) -> int:
+        coverage = np.average(
+            list(map(lambda result: getattr(result, path_to_coverage), results))
         )
+        return coverage
+
+    average_result = avg_coverage(
```
Comment (Parzival-05) [inline AIAgent/ml/training/validation.py:92]: Может, вынести в StatisticsCollector? И "avg_by_attr"?

File: AIAgent/ml/training/validation.py [hunk 5]
```diff
@@ -78,38 +82,42 @@ def validate_coverage(
...
-        ),
-    )
-    average_result = np.average(
-        list(
-            map(
-                lambda map_result: map_result.game_result.actual_coverage_percent,
-                all_results,
-            )
+
+    def avg_coverage(results, path_to_coverage: str) -> int:
```
Comment (Parzival-05) [inline AIAgent/ml/training/validation.py:86]: Может, вынести в StatisticsCollector? И "avg_by_attr"?

File: AIAgent/ml/training/validation.py (hunk 1 above)
Comment (Anya497) [reply AIAgent/ml/training/validation.py:71]: Сейчас в файле сохраняется только одна эпоха, а этот файл каждую эпоху логируется mlflow. Я думаю, пока можно так оставить, а потом вообще отказаться от этих табличек, потому что с ними неудобно работать.

File: AIAgent/ml/training/validation.py (hunk 2 above)
Comment (Anya497) [reply AIAgent/ml/training/validation.py:88]: Да. Это в другой ветке поправлено. Наверное, лучше ее сразу влить

File: AIAgent/ml/training/validation.py (hunk 3 above)
Comment (Anya497) [reply AIAgent/ml/training/validation.py:103]: Да, это нужно

File: AIAgent/ml/training/validation.py (hunk 2 above)
Comment (Anya497) [reply AIAgent/ml/training/validation.py:88]: А, нет, это же results внутри функции. Там в другом месте надо было заменить, чуть ниже.

File: AIAgent/common/classes.py [hunk 6]
```diff
@@ -3,7 +3,7 @@
 from dataclasses_json import dataclass_json
 from typing import Optional, TypeAlias
 
-from common.game import GameMap
+from common.game import GameMap, GameMap2SVM
```
Comment (gsvgit) [inline AIAgent/common/classes.py:6]: `GameMap2SVM` looks strange. We need new clear name.

File: AIAgent/common/game.py [hunk 7]
```diff
@@ -74,7 +74,12 @@ class GameMap:
     NameOfObjectToCover: str
     DefaultSearcher: str
     MapName: str
-    SVMInfo: SVMInfoT = field(default=None, metadata=config(exclude=lambda x: True))
+
+
+@dataclass(slots=True)
+class GameMap2SVM:
```
Comment (gsvgit) [inline AIAgent/common/game.py:80]: `BindedGameMap` ?

File: AIAgent/ml/play_game.py [hunk 8]
```diff
@@ -142,43 +141,36 @@ def play_map_with_timeout(
 def play_game(
     with_predictor: Predictor,
     max_steps: int,
-    maps: list[GameMap],
+    game_map2svm: GameMap2SVM,
```
Comment (gsvgit) [inline AIAgent/ml/play_game.py:144]: Variable name looks strange too.

File: AIAgent/ml/play_game.py [hunk 9]
```diff
@@ -142,43 +141,36 @@ def play_map_with_timeout(
...
-            logging.warning(log_message)
-            list_of_failed_maps_with_errors.append((game_map, error))
-    if list_of_failed_maps_with_errors:
+        logging.info(
+            f"<{with_predictor.name()}> finished map {game_map2svm.GameMap.MapName} "
+            f"in {game_result.steps_count} steps, {time} seconds, "
+            f"actual coverage: {game_result.actual_coverage_percent:.2f}"
+        )
+        map2result = Map2Result(game_map2svm, game_result)
+    except (FunctionTimedOut, Exception) as error:
```
Comment (gsvgit) [inline AIAgent/ml/play_game.py:166]: Why not
```
except Error1:
  print("Variable x is not defined")
except Error2:
  ...
```

File: AIAgent/run_training.py [hunk 10]
```diff
@@ -114,63 +108,82 @@ def run_training(
...
+    model_init: Callable[[Any], nn.Module],
     epochs: int,
-    models_path: str,
-    server_count: int,
+    val_config: ValidationConfig,
 ):
     config = TrialSettings(
-        lr=0.0003,  # trial.suggest_float("lr", 1e-7, 1e-3),
-        batch_size=109,  # trial.suggest_int("batch_size", 8, 1800),
+        lr=trial.suggest_float("lr", 1e-7, 1e-3),
```
Comment (gsvgit) [inline AIAgent/run_training.py:172]: Can we move all these parameters to the config?

File: AIAgent/common/classes.py (hunk 6 above)
Comment (Anya497) [reply AIAgent/common/classes.py:6]: Почему? Эта штука как раз сопоставляет информацию о карте символьной машине. Еще у нас есть структура `Map2Result`, например, которая несет в себе похожий смысл.

File: AIAgent/ml/play_game.py (hunk 9 above)
Comment (Anya497) [reply AIAgent/ml/play_game.py:166]: Следующим ПРом волью. [Вот коммит.](https://github.com/gsvgit/PySymGym/commit/3f7573a7cb1e5f511cb717f2a0afe13556fa5bf8 'Refactor exceptions handling.')

File: AIAgent/run_training.py (hunk 10 above)
Comment (Anya497) [reply AIAgent/run_training.py:172]: Нверное, лучше так сделать. Думаю, отдельным ПРом. [Я записала](https://github.com/gsvgit/PySymGym/issues/77)

**Review by gsvgit — APPROVED** [review]

**Review by Parzival-05 — APPROVED** [review]

File: AIAgent/ml/play_game.py (hunk 9 above)
Comment (Anya497) [reply AIAgent/ml/play_game.py]: А нет, все-таки в этом будет

File: AIAgent/ml/play_game.py [hunk 11]
```diff
@@ -163,14 +163,13 @@ def play_game(
...
-        logging.warning(log_message)
-        FeatureConfig.SAVE_IF_FAIL_OR_TIMEOUT.save_model(
-            with_predictor.model(), with_name=f"{with_predictor.name()}"
-        )
-        map2result = Map2Result(game_map2svm, None)
+    except FunctionTimedOut as error:
+        log_message = f"<{with_predictor.name()}> timeouted on map {game_map2svm.GameMap.MapName} with {error.timedOutAfter}s"
+    except Exception:
+        log_message = f"<{with_predictor.name()}> failed on map {game_map2svm.GameMap.MapName}:\n{traceback.format_exc()}"
+    logging.warning(log_message)
```
Comment (Parzival-05) [inline AIAgent/ml/play_game.py:170]: `log_message` может быть не определен, когда не выбросилось исключение
